### PR TITLE
Redirect classic url to new format url

### LINKED ISSUE #45: Maven Central Badge has wrong format url
Thanks for creating an issue! Please fill out this form so we can be
sure to have all the information we need, and to minimize back and forth.

* **What are you trying to do?**
Copy the generated Maven Central Badge Markdown link and use it in README.md
(link format = http://search.maven.org/#search%7Cga%7C1%7Cg%3A%22${groupId}%22%20a%3A%22${artifactId}%22, but in new version of search.maven.org the link format should be https://search.maven.org/search?q=g:%22${groupId}%22%20AND%20a:%22${artifactId}%22&core=gav)

* **What feature or behavior is this required for?**
Click the Maven Central Badge to open the search.maven.org and navigate to the specified artifact.

* **How could we solve this issue? (Not knowing is okay!)**
Update the link format to the right one, I will contribute a PR to fix it.

* **Anything else?**
none

### PR DESCRIPTION
**Pull request by ian4hu**
Some external resource (like Github Readme etc.) reference artifacts with classic url format now always fallback to the home page. We need to detect the old format url and then redirect it to new one.

This pull request makes the following changes:
* Add a ClassicModule to provide a ClassicUrlGuard
* Add a ClassicUrlGuard to detect and redirect the classic url format to new one
* Register the ClassicModule into AppModule

(If there are changes to the UI, please include a screenshot so we
know what to look for!) no

(If there are changes to user behavior in general, please make sure to
update the docs, as well) no

It relates to the following issue #s:
* Fixes #45 #43

### CODE REVIEW

**Comment by ian4hu** [comment]
Handle classic search
old format `#search|gav|1|g:{groupId} a:{artifactId} v:{version}` to new format `/search?q=g:{groupId} AND a:{artifactId} AND v:{version}&core={gav or empty}`

and classic artifact detail
old format `#artifactdetails|{groupId}|{artifactId}|{version}|{packaging}`  to `new format /artifact/{groupId}/{artifactId}/{version}/{packaging}`

**Comment by SirMaster** [comment]
@ian4hu really cool stuff thank you for contributing back (again)! We hope to look at this a bit further later today.

**Comment by ian4hu** [comment]
@SirMaster I have added the Sonatype copyright header to new files.

**Comment by oliviercailloux** [comment]
Does that mean there’s an intent of restoring compatibility with the old linking behavior? That would really be awesome as I have links in the old format scattered everywhere.

I understand that there has never been any explicit commitment to have stable links, but nevertheless I had expected (perhaps unduly) that the implicit contract was that http://search.maven.org/ would not break those links, given that many people use them, and given that there has never been (AFAIK) any explicit warning against using those links or any other method of linking directly to an artifact that would have been officially recommended, and given that there is an implicit [convention](https://www.w3.org/Provider/Style/URI) (or best practice) on the web that urls shouldn’t change (I admit this is debatable, of course, as in practice urls do change).

Long story short: if you can ensure compatibility with the old links, at least for some time in the future, that would be awesome, not only for me but, I suspect, for many users.

**Comment by DarthHater** [comment]
@ian4hu we are going through the paces to import this and get it out, @SirMaster found some more cases to get working, and he's addressing them on an internal PR at the moment. I'll let you know when it gets merged!

**Comment by DarthHater** [comment]
@ian4hu this is now merged and out in master, the merge ended up going wacky ( @SirMaster got the authorship on it), but rest assured we all know you did it, and we are super stoked! Thanks again for contributing, you rock!

**Comment by oliviercailloux** [comment]
Thanks to all, very good to see that previous url behavior are restored.
